### ISSUE TITLE
Failure to account for None-returning generators in StopIteration.args

### ISSUE THREAD

**Issue by dvarrazzo** [python/mypy#18073]
In psycopg 3 we have generators returning a final result, which sometimes is None, sometimes a different object. We also have some consumers of these generators, which wait for I/O completion according to the values yielded by the generators and finally return the final generator result.

Something similar can be represented with the following code:

```python
from typing import Any, Generator, TypeAlias, TypeVar

RV = TypeVar("RV")

PQGen: TypeAlias = Generator[Any, Any, RV]


def str_gen() -> PQGen[str]:
    yield 10
    return "abc"


def none_gen() -> PQGen[None]:
    yield 10
    return None


def wait(gen: PQGen[RV]) -> RV:
    try:
        while True:
            next(gen)
    except StopIteration as ex:
        rv: RV = ex.args[0] if ex.args else None
        return rv


print(f"{wait(str_gen())=}")
print(f"{wait(none_gen())=}")
```

If a generator returns `None`, the `StopIteration.args` tuple will empty, which is accounted for in the `wait()`'s return statement (removing the check will cause an IndexError on `wait(none_gen())`). However, since Mypy 1.12, the return statement in the wait returns an error:

```
bug_mypy.py:24: error: Incompatible types in assignment (expression has type "Any | None", variable has type "RV")  [assignment]
```

This wasn't reported as error in Mypy 1.11

**Comment by brianschubert** [comment]
What Jelle said. Do note that there’s no connection in the type system between `RV` and the type of `ex`. Mypy has no way of knowing that "`ex.args` is empty at runtime" means "`RV` must be `None`", which is what that assignment assumes. Usually, in order to inject that sort of runtime knowledge into the type system, you’d need to use a `cast` or a narrowing construct (asserts, type guards, etc.).

Also, you may find it easier to use [`StopIteration.value`](https://docs.python.org/3/library/exceptions.html#StopIteration.value) (which is meant for this purpose) instead of trying to parse `BaseException.args`. `value` has a type of `Any`, so simply returning `ex.value` will type check.

**Comment by dvarrazzo** [comment]
Yes, the primer in #17427 indeed shows the changes in psycopg check.

Observation: If expr is Any, isn't "exprt if cond else None` an Any too?

However, thank you very much for pointing out the use of `StopIteration.value`, @brianschubert! I missed it, and it seems we can use it instead of `args[0]` to work around the problem. I think this issue can be closed.

**Comment by JelleZijlstra** [comment]
> Observation: If expr is Any, isn't "expr if cond else None" an Any too?

No, it is `Any | None`, which is a different type. For example, accessing arbitrary attributes is OK on a value of type `Any`, but not on a value of type `Any | None`.
